wallet/src/account/utxo_selector/mod.rs

```diff
@@ -0,0 +1,695 @@
+                    result.clear();
+                    result.add_input(group, pay_fees)?;
+                } else {
+                    // No UTXO above target, nothing to do.
+                    return Err(UtxoSelectorError::MaxWeightExceeded);
+                }
+            }
+        }
+    }
+
+    Ok(result)
+}
+
+/// Find a subset of the OutputGroups that is at least as large as, but as close as possible to, the
+/// target amount; solve subset sum.
+fn approximate_best_subset(
+    rng: &mut impl Rng,
+    groups: &[OutputGroup],
+    total_lower: Amount,
+    target_value: Amount,
+    best_solution_selected: &mut [bool],
+    best_solution_total: &mut Amount,
+    pay_fees: PayFee,
+) {
+    let mut current_solution_included: Vec<bool> = vec![false; groups.len()];
+    // initial best is including all groups
+    best_solution_selected.iter_mut().for_each(|elem| *elem = true);
+    *best_solution_total = total_lower;
+
+    for _ in 0..1000 {
+        // reset selection to all false
+        current_solution_included.iter_mut().for_each(|elem| *elem = false);
+        let mut total = Amount::ZERO;
+        let mut target_reached = false;
+
+        for n_pass in 0..2 {
+            for (i, group) in groups.iter().enumerate() {
+                //The solver here uses a randomized algorithm,
+                //the randomness serves no real security purpose but is just
+                //needed to prevent degenerate behavior and it is important
+                //that the rng is fast. We do not use a constant random sequence,
+                //because there may be some privacy improvement by making
+                //the selection random.
+                let rand_bool = if n_pass == 0 {
+                    rng.gen::<bool>()
+                } else {
+                    !current_solution_included[i]
+                };
+
+                if rand_bool {
+                    total = (total + group.get_effective_value(pay_fees))
+                        .expect("total sum has been checked to not overflow");
+                    current_solution_included[i] = true;
+
+                    if total >= target_value {
+                        target_reached = true;
+
+                        if total < *best_solution_total {
+                            *best_solution_total = total;
+                            best_solution_selected.clone_from_slice(&current_solution_included);
+                        }
+
+                        total = (total - group.get_effective_value(pay_fees))
+                            .expect("total sum has been checked to not overflow");
+                        current_solution_included[i] = false;
+                    }
+                }
+            }
+
+            if target_reached {
+                break;
+            }
+        }
+    }
+}
+
+/// This is the Branch and Bound Coin Selection algorithm designed by Murch. It searches for an input
```

Comment:
Maybe we should update the license notice, since some of the code comes from Bitcoin Core?

Reply:
I added the fallowing section to the license notice of this file:
```
// Original code copyright (c) 2011-2022 The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```